core/shared/src/main/scala/zio/prelude/fx/ZPure.scala

```diff
@@ -464,38 +492,42 @@ sealed trait ZPure[-S1, +S2, -R, +E, +A] { self =>
   final def run(s: S1)(implicit ev1: Any <:< R, ev2: E <:< Nothing): (S2, A) =
     runEither(s).fold(ev2, identity)
 
+  final def runEither(s: S1)(implicit ev: Any <:< R, ev1: CanFail[E]): Either[E, (S2, A)] =
+    runEitherCause(s).fold(cause => Left(cause.failure), Right(_))
+
   /**
    * Runs this computation with the specified initial state, returning either a
    * failure or the updated state and the result
    */
-  final def runEither(s: S1)(implicit ev: Any <:< R, ev1: CanFail[E]): Either[E, (S2, A)] = {
-    val _                                                   = ev
-    val stack: Stack[Any => ZPure[Any, Any, Any, Any, Any]] = Stack()
-    val environments: Stack[AnyRef]                         = Stack()
-    var s0: Any                                             = s
-    var a: Any                                              = null
-    var failed                                              = false
-    var curZPure: ZPure[Any, Any, Any, Any, Any]            = self.asInstanceOf[ZPure[Any, Any, Any, Any, Any]]
+  final def runEitherCause(s: S1)(implicit ev: Any <:< R, ev1: CanFail[E]): Either[Cause[E], (S2, A)] = {
```

Comment:
The log is "eaten", right? That's ok if we expose "get log". In fact, maybe through "get state", we could further simplify this to just `Either[Cause[E], A]`, so if you want the state or the log, you have to get it and bundle it into the result.

Reply:
I like that.